File: snippets/lqr.py
"""Function for computing the infinite horizon LQR."""

import numpy as np
import scipy as sp


def lqr(A, B, Q, R, N):
    """
    Solves for the optimal discrete linear-quadratic regulator (LQR).

    Args:
        A: System matrix, states x states.
        B: Input matrix, states x inputs.
        Q: State cost matrix, states x states.
        R: Control effort cost matrix, inputs x inputs.
        N: Cross weight matrix, states x inputs.

    Returns:
        Controller gain matrix, inputs x states.
    """
    P = sp.linalg.solve_discrete_are(a=A, b=B, q=Q, r=R, s=N)
    return np.linalg.solve(B.T @ P @ B + R, B.T @ P @ A + N.T)

File: snippets/LQR.cpp
#include <Eigen/Cholesky>
#include <Eigen/Core>

#include "DARE.hpp"

template <int States, int Inputs>
Eigen::Matrix<double, Inputs, States> LQR(
    const Eigen::Matrix<double, States, States>& A,
    const Eigen::Matrix<double, States, Inputs>& B,
    const Eigen::Matrix<double, States, States>& Q,
    const Eigen::Matrix<double, Inputs, Inputs>& R,
    const Eigen::Matrix<double, States, Inputs>& N) {
  using StateMatrix = Eigen::Matrix<double, States, States>;

  auto R_llt = R.llt();
  StateMatrix A_2 = A - B * R_llt.solve(N.transpose());
  StateMatrix Q_2 = Q - N * R_llt.solve(N.transpose());

  StateMatrix P = DARE<States, Inputs>(A_2, B, Q_2, R);

  return (B.transpose() * P * B + R)
      .llt()
      .solve(B.transpose() * P * A + N.transpose());
}
